### ISSUE TITLE
[android] .so fails to load: not accessible

### ISSUE THREAD

**Issue by eddiegalvez** [rsn8887/capsimg#1]
library "/storage...RetroArch/system/capsimg.so" needed or dlopened by "/data/data/com.retroarch/cores/puae_libretro_android.so" is not accessible for the namespace: [name="classloader-namespace", ld_library_paths="", default_library_paths="/data/app/com.retroarch-1/lib/arm64:/data/app/...base.apk!/lib/arm64-v8a", permitted_paths="/data:/mnt/expand:/data/data/com.retroarch"]

**Comment by rsn8887** [comment]
Are you sure you installed the correct .so version for your device? There are many different ones in the archive: https://github.com/rsn8887/capsimg/releases/download/1.0/Capsimg_for_Retroarch.zip

You have to pick the correct one and copy it to Retroarch system folder. It looks like you need the one from the folder `arm64-v8a`

Edit: This could be a permission issue. Do you have root permissions on your device by any chance?

**Comment by eddiegalvez** [comment]
Yes. GPD XD+ with CleanRom and rooted (unless I messed up)

**Comment by eddiegalvez** [comment]
Note that I removed the .so from the retroarch system folder and instead placed it in the /data/com.retroarch folder (that is, inside the app folder itself) and now the log doesn't have that error anymore (but I have no valid test since I'm trying to use PUAE via retroarch and all I get still is the amiga "insert disk" image).

**Comment by turboboy215** [comment]
I am having the exact same problem. It looks like the arm64-v8a version is broken. Have you confirmed this works? I can't find an ADF version of one of the games I'm trying to run.

**Comment by rsn8887** [comment]
Me? No. I don't have such a device. I don't really use Android myself. I have an old Nvidia Shield Portable and I tested on that (armeabi-v7a). I took the code from lubomyr's UAE4Arm Android repo.
https://github.com/lubomyr/uae4arm/tree/master/plugins/capsimage

Does the lib work with UAE4Arm on your device? I suspect there's just some incompatibility between Retroarch and the lib. Maybe some compiler flags etc.

If you want to run a game, you can always use TOSEC adf version or WHDLoad for the game you want to run, there are so many options.

**Comment by rsn8887** [comment]
This might help, I am not sure:
https://source.android.com/devices/tech/config/namespaces_libraries

**Comment by rsn8887** [comment]
And this, looks like Retroarch is built with a low target Android version so this behavior breaks on later Android maybe. https://android-developers.googleblog.com/2016/06/improving-stability-with-private-cc.html

**Comment by rsn8887** [comment]
You could try to put the .so in /data/data/com.retroarch like the error suggests, or did you already?

**Comment by Panderner** [comment]
> You could try to put the .so in /data/data/com.retroarch like the error suggests, or did you already?

It needs to root to access /data folder

**Comment by asdf288** [comment]
So if I understood this correctly, the problem is that Android only allows an app to run libraries that are located in its own app or data folder, for security reasons I assume? That would mean it's impossible for a non-root user to download and install the library, because he has no access to these folders. The library would have to be downloaded by Retroarch itself.

Until this is solved, here are the instructions to get it working, root access is required. Tested on Android 11.

- Install FX file explorer with root plugin
- Create folder "/data/user/0/com.retroarch.aarch64/system"
- Copy contents of your old system folder to this new folder, including the capsimg.so file
- Edit retroarch.cfg (/sdcard/Android/data/com.retroarch.aarch64/files/retroarch.cfg) and set "system_directory" to the new system folder above (this can't be done from the Retroarch menu, because the default user cannot see these folders in the built-in file browser)

If the game fails to load and you're stuck at the "insert disk" screen, check the log file for errors (for example, I had installed the arm64 version of the library, but the PUAE core was 32 bit, so it failed to load).

**Comment by tharubberchicken** [comment]
Further to this, I have discovered a workaround (based off asdf288's) so you do not have to copy your _entire_ system folder to **/data/data/com.retroarch.aarch64/system** if like me your system folder is quite big. Here is what you do:

1) create a file named **PUAE.cfg** under **/storage/emulated/0/RetroArch/config/PUAE** or wherever your core configs are stored – it is okay if this file already exists!

2) add the following line to the new or existing **PUAE.cfg** file and save it:
`system_directory = "data/data/com.retroarch.aarch64/system"`
I used **data/user/0/com.retroarch.aarch64/system** which worked the same for me.

3) transfer ONLY your Amiga/PUAE system files to this new system directory, namely:
capsimg.so
kick*.*
WHDLoad.*
uae_data (folder)

A bit of a fiddle, but it works a treat.

**Comment by Korvenwin** [comment]
To do this, is it necesary to root the device?

**Comment by tharubberchicken** [comment]
Yes, unfortunately. At least for more recent versions of Android, 11 and up, maybe? This was a while ago when I did this and I still am doing it this way. If I remember right, using other directories outside of the RetroArch's data folder didn't work. I can relook into it.

**Comment by maddie480** [comment]
For people like me stumbling on this issue on an unrooted device: a workaround has been implemented on PUAE's side (https://github.com/libretro/libretro-uae/pull/688). You can now install `capsimg.so` as if it was a core, so RetroArch does the copying to `/data/user/0/com.retroarch/cores` itself, and PUAE will pick it up there.

**Comment by tharubberchicken** [comment]
This is great. 👍🏻
